lib/ecto/association.ex

```diff
@@ -235,8 +236,10 @@ defmodule Ecto.Association do
     # for the final WHERE clause with values.
     {_, query, _, dest_out_key} = Enum.reduce(joins, {source, query, counter, source.out_key}, fn curr_rel, {prev_rel, query, counter, _} ->
       related_queryable = curr_rel.schema
-
-      next = join(query, :inner, [{src, counter}], dest in ^related_queryable, on: field(src, ^prev_rel.out_key) == field(dest, ^curr_rel.in_key))
+      next = query
+        # join on the foreign key
+        |> join(:inner, [{src, counter}], dest in ^related_queryable, on: ^on_fields(prev_rel.out_key, curr_rel.in_key))
```

Comment:
Unfortunately I don't think this is correct. The `src` is not always going to be the one immediately before the `dest`. You need to pass the `counter` to `on_fields` in order to build the proper source. Using `dynamic` is not ideal either. We will need to find ways to optimize it but that can be a later concern.